packages/accounts-controller/src/AccountsController.ts

```diff
@@ -0,0 +1,580 @@
+        this.setSelectedAccount(this.listAccounts()[0].id);
+      } else if (
+        updatedKeyringAddresses.length < previousAccounts.length &&
+        overlaps.length > 0 &&
+        !this.getAccount(this.state.internalAccounts.selectedAccount)
+      ) {
+        this.#handleSelectedAccountRemoved();
+      }
+    }
+  }
+
+  /**
+   * Handles the change in SnapControllerState by updating the metadata of accounts that have a snap enabled.
+   *
+   * @param snapState - The new SnapControllerState.
+   * @returns A Promise that resolves when the update is complete.
+   */
+  async #handleOnSnapStateChange(
+    snapState: SnapControllerState,
+  ): Promise<void> {
+    // only check if snaps changed in status
+    const { snaps } = snapState;
+    const accounts = this.listAccounts().filter(
+      (account) => account.metadata.snap,
+    );
+
+    this.update((currentState: AccountsControllerState) => {
+      accounts.forEach((account) => {
+        const currentAccount =
+          currentState.internalAccounts.accounts[account.id];
+        if (currentAccount?.metadata?.snap) {
+          currentAccount.metadata.snap.enabled =
+            snaps[currentAccount.metadata.snap.id].enabled &&
+            !snaps[currentAccount.metadata.snap.id].blocked;
+        }
+      });
+    });
+  }
+
+  /**
+   * Handles the event when a new account is added to the keyring.
+   *
+   * @param updatedKeyringAddresses - An array of updated keyring addresses.
+   * @param previousAccounts - An array of previous internal accounts.
+   */
+  #handleNewAccountAdded(
+    updatedKeyringAddresses: string[],
+    previousAccounts: InternalAccount[],
+  ) {
+    const [newAddress] = updatedKeyringAddresses.filter(
+      (address) =>
+        !previousAccounts.find(
+          (account) => account.address.toLowerCase() === address.toLowerCase(),
+        ),
+    );
+
+    const [newAccount] = this.listAccounts().filter(
+      (account) => account.address.toLowerCase() === newAddress.toLowerCase(),
+    );
+
+    this.setSelectedAccount(newAccount.id);
+  }
+}
+
+/**
+ * Returns the name of the keyring type.
+ *
+ * @param keyringType - The type of the keyring.
+ * @returns The name of the keyring type.
+ */
+export function keyringTypeToName(keyringType: string): string {
+  switch (keyringType) {
+    case 'Simple Key Pair': {
+      return 'Account';
+    }
+    case 'HD Key Tree': {
+      return 'Account';
+    }
+    case 'Trezor Hardware': {
+      return 'Trezor';
+    }
+    case 'Ledger Hardware': {
+      return 'Ledger';
+    }
+    case 'Lattice Hardware': {
+      return 'Lattice';
+    }
+    case 'QR Hardware Wallet Device': {
+      return 'QR';
+    }
+    case 'Snap Keyring': {
+      return 'Snap Account';
+    }
+    case 'Custody': {
+      return 'Custody';
+    }
+    default: {
+      console.warn(`Unknown keyring ${keyringType}`);
+      return 'Account';
+    }
```

Comment:
Is it safe to return the same string for both known and unknown keyrings?

Reply:
It should really never reach this path, but i agree, the name can be `Unknown` instead of `Account`

Reply:
Just curious, what would happen if eventually we reach this branch and get an `Unknown` as keyring?
(My real question is, does it make sense to throw an error here instead?)

Reply:
It would make sense to throw an error, because we don't really want an Unknown account name.